# Run in this env.: /medstore/Development/nanopore_HBV/daniel/micromamba/envs/snakemake_hbv
# Dry run with: snakemake -np
# Run with e.g.: snakemake --cores 4

FILES = glob_wildcards("data/{sample}.fastq.gz")    # get all fastq files

NAMES = FILES.sample    # extract the sample names from the files

rule all:    # specify the final output files
    input:
        expand("fastp/{sample}_filt.fastq.gz", sample=NAMES),
        expand("minimap2/{sample}.sam", sample=NAMES),
        expand("samtools/{sample}.bam", sample=NAMES)

rule fastp:    # filter reads
    input:
        fastq = "data/{sample}.fastq.gz"
    output: 
        fastq = "fastp/{sample}_filt.fastq.gz",    # output filtered read
        report = "fastp/report/{sample}_filt.html"       # output fastp report
    log:
        "logs/fastp/{sample}.log"
    shell:
        "fastp -i {input.fastq} -o {output.fastq} -h {output.report}"

rule minimap2:    # map reads to reference genomes.
    input:
        ref = "reference_genomes/ref_a.fa",    # only one at the moment, the others will be added
        fastq = "fastp/{sample}_filt.fastq.gz"    # filtered reads
    output:
        sam = "minimap2/{sample}.sam"
    shell:
        "minimap2 -ax map-ont {input.ref} {input.fastq} > {output.sam}"

rule samtools:
    input:
        sam = "minimap2/{sample}.sam"
    output:
        bam = "samtools/{sample}.bam"
    shell:
        "samtools view -bS {input.sam} | samtools sort -o {output.bam}"